src/lib/block_if.c

```diff
@@ -221,22 +228,46 @@ block_delete(struct blockif_ctxt *bc, off_t offset, off_t len)
 		errno = EOPNOTSUPP;
 	else if (bc->bc_rdonly)
 		errno = EROFS;
-	if (bc->bc_fd >= 0) {
-		if (bc->bc_ischr) {
-#ifdef __FreeBSD__
-			ret = ioctl(bc->bc_fd, DIOCGDELETE, arg);
-#else
-			errno = EOPNOTSUPP;
-#endif
-		} else
-			errno = EOPNOTSUPP;
 #ifdef HAVE_OCAML_QCOW
-	} else if (bc->bc_mbh >= 0) {
+	if (bc->bc_mbh >= 0) {
 		ret = mirage_block_delete(bc->bc_mbh, offset, len);
+		goto out;
+	}
 #endif
-	} else
-		abort();
-
+#ifdef __FreeBSD__
+	if ((bc->bc_fd >= 0) && (bc->bc_ischr)) {
+		ret = ioctl(bc->bc_fd, DIOCGDELETE, arg);
+		errno = EOPNOTSUPP;
+		goto out;
+	}
+#elif __APPLE__
+	if (bc->bc_fd >= 0) {
+		/* PUNCHHOLE lengths and offsets have to be aligned so explicitly write zeroes
+		   into the unaligned parts at the beginning and the end. This wouldn't be necessary
+		   if the host and guest had the same sector size */
+		offset_unaligned_by = arg.fp_offset - (arg.fp_offset / bc->bc_delete_alignment * bc->bc_delete_alignment);
```

Comment:
Could you encapsulate `arg.fp_offset / bc->bc_delete_alignment * bc->bc_delete_alignment` into a `ROUND_DOWN(arg.fp_offset, bc->bc_delete_alignment)` please. You might be able to steal `src/lib/firmware/kexec.c:#define ALIGNDOWN(x, a) (-(a) & (x))`? (I assume `bc_delete_alignment` is a power of two?)